Tracing support for Kafka Mirror Maker and Kafka Connect

scholzj opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This PR adds support for Jaeger tracing in Kafka Mirror Maker and Kafka Connect. It let's users to enable tracing which will:
* Configure the interceptors in Mirror Maker consumers and producers and in Connect consuemrs and producers
* Enable tracing agent which initializes Jaeger tracing client at startup

The configuration of the Jaeger client should be done from the template environment variables (since it is pretty much unbounded list of configuration variables which might change over time). This is currently not present in MIrror Maker and Connect, but is being worked on in different work stream.

The OpenTracing Kafka libraries are dependent on Kafka versions. Therefore we need to download different versions of the library for different version of Kafka. As a result, 

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/tracing/JaegerTracing.java:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2018, Strimzi authors.
```
New file. Better 2019.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/tracing/Tracing.java:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2018, Strimzi authors.
```
Ditto as above.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/tracing/Tracing.java:
```diff
@@ -0,0 +1,53 @@
+            "Currently the supported types are `route`, `loadbalancer`, and `nodeport`. \n\n" +
+            "* `route` type uses OpenShift Routes to expose Kafka." +
+            "* `loadbalancer` type uses LoadBalancer type services to expose Kafka." +
+            "* `nodeport` type uses NodePort type services to expose Kafka.")
```
I guess this description is coming from some copy/paste :)

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:
```diff
@@ -138,7 +141,14 @@ public static KafkaConnectCluster fromCrd(KafkaConnect kafkaConnect, KafkaVersio
+        kafkaConnect.tracing = spec.getTracing();
+
+        KafkaConnectConfiguration config = new KafkaConnectConfiguration(spec.getConfig().entrySet());
+        if (kafkaConnect.tracing != null && "jaeger".equals(kafkaConnect.tracing.getType()))   {
```
Can we use the TYPE_JAEGER constant instead of hard coded string?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java:
```diff
@@ -400,13 +405,33 @@ public Deployment generateDeployment(Map<String, String> annotations, boolean is
+    private KafkaMirrorMakerConsumerConfiguration getConsumerConfiguration()    {
+        KafkaMirrorMakerConsumerConfiguration config = new KafkaMirrorMakerConsumerConfiguration(consumer.getConfig().entrySet());
+
+        if (tracing != null && "jaeger".equals(tracing.getType())) {
```
Ditto as above

ppatierno commented on docker-images/kafka/kafka-thirdparty-libs/2.1.x/pom.xml:
```diff
@@ -3,7 +3,7 @@
     <groupId>io.strimzi</groupId>
-    <version>0.13.0-SNAPSHOT</version>
+    <version>1.0.0-SNAPSHOT</version>
     <modelVersion>4.0.0</modelVersion>
```
Why this jump to 1.0.0?

scholzj replied:
This was originally setup to have the same version as the Strimzi project. But it is not part of the main Maven build it is not updated automatically. It also doesn't make sense for this to have the same version as Strimzi it self when it is not plugged into the main Maven project. So to avoid confusion, I changed the version to _neutral_ `1.0.0-SNAPSHOT`. Do you have any better idea what version should it have?

ppatierno commented on docker-images/kafka/scripts/kafka_connect_run.sh:
```diff
@@ -33,7 +33,12 @@ export LOG_DIR="$KAFKA_HOME"
+  export KAFKA_OPTS="${KAFKA_OPTS} -javaagent:$(ls $KAFKA_HOME/libs/jmx_prometheus_javaagent*.jar)=9404:$KAFKA_HOME/custom-config/metrics-config.yml"
+fi
+
+# enabling Tracing agant (initializes Jaeger tracing) as Java agent
```
Typo "agant"

ppatierno commented on docker-images/kafka/scripts/kafka_mirror_maker_run.sh:
```diff
@@ -56,6 +56,11 @@ if [ "$KAFKA_MIRRORMAKER_METRICS_ENABLED" = "true" ]; then
   export KAFKA_OPTS="$KAFKA_OPTS -javaagent:$(ls $KAFKA_HOME/libs/jmx_prometheus_javaagent*.jar)=9404:$KAFKA_HOME/custom-config/metrics-config.yml"
 fi
 
+# enabling Tracing agant (initializes Jaeger tracing) as Java agent
```
Ditto as above

scholzj commented:
@ppatierno I rebased this against master, updated Jaeger to 1.0.0 and did some finnishing touches. Could you please review this once more? (there are no significant changes since last time, but it is the final PR now).
